### PR TITLE
First service node test

### PR DESCRIPTION
**Pull request by msgmaxim**
- Added a first service node test that checks balance after a registration transaction
- Added a helper function to calculate unlocked balance in tests

### CODE REVIEW

**Comment by jcktm** [comment]
This has some things in it that could change consensus that I haven't tested. Maybe take out the service node list refactoring changes, they can only cause problems. Functions that were added to core are unused also. Sorry maxim I sent you a WIP branch

**Comment by jcktm** [comment]
This test doesn't look finished, it tests that a registration tx can be constructed with the test framework... but a registration tx is just a regular tx, so all this checks in core:: is that a regular tx can be submitted to the blockchain.

Ideally this should at least test e.g. that alice is receiving payouts, and the test should be renamed to check_basic_registration or something.

File: tests/core_tests/chaingen.cpp [hunk 1]
```diff
@@ -582,13 +587,15 @@ void fill_tx_sources_and_destinations(const std::vector<test_event_entry>& event
...
+  uint64_t cash_back = get_inputs_amount(sources) - (amount + fee);
+  if (0 < cash_back)
   {
-    if (!fill_tx_destination(de_change, from, cache_back))
+    if (!fill_tx_destination(de_change, from, cash_back))
       throw std::runtime_error("couldn't fill transaction cache back destination");
     destinations.push_back(de_change);
   }
+
+  if (change_amount) *change_amount = (cash_back > 0) ? cash_back : 0;
```
Comment (jcktm) [inline tests/core_tests/chaingen.cpp]: boost::optional is less bug prone than null pointers @msgmaxim I actually was reminded about this when reading the rust book. I see what you mean now about compile-time reference checks, very cool

**Comment by msgmaxim** [comment]
You are right, the test was not finished. Just yesterday I've modified the make_registration_tx to do the right thing. Should come in the new pull request.

File: tests/core_tests/chaingen.cpp (hunk 1 above)
Comment (msgmaxim) [reply tests/core_tests/chaingen.cpp]: Nice! :)

**Pull request merged**